cmd/prow-job-dispatcher/main.go

```diff
@@ -61,9 +83,248 @@ func (o *options) complete(secrets *sets.String) error {
+	}
+
+	cv.mutex.Unlock()
+	return cluster
+}
+
+func (cv *clusterVolume) addToVolume(cloudProvider, cluster string, jobBase prowconfig.JobBase, path string, config *dispatcher.Config, jobVolumes map[string]float64) {
+	determinedCluster, canBeRelocated := config.DetermineClusterForJob(jobBase, path)
+	if cluster == string(determinedCluster) || canBeRelocated {
+		cv.clusterVolumeMap[cloudProvider][cluster] = cv.clusterVolumeMap[cloudProvider][cluster] + jobVolumes[jobBase.Name]
+	} else if determinedCloudProvider, ok := config.IsInBuildFarm(determinedCluster); ok {
+		cv.clusterVolumeMap[string(determinedCloudProvider)][string(determinedCluster)] = cv.clusterVolumeMap[string(determinedCloudProvider)][string(determinedCluster)] + jobVolumes[jobBase.Name]
+	}
+}
+
+// dispatchJobConfig dispatches the jobs defined in a Prow jon config
+func (cv *clusterVolume) dispatchJobConfig(jc *prowconfig.JobConfig, path string, config *dispatcher.Config, jobVolumes map[string]float64) string {
+	cloudProvidersForE2ETests := getCloudProvidersForE2ETests(jc)
+	var cluster string
+	if cloudProvidersForE2ETests.Len() == 1 {
+		cloudProvider, _ := cloudProvidersForE2ETests.PopAny()
+		cluster = cv.findClusterForJobConfig(cloudProvider, jc, path, config, jobVolumes)
+	} else {
+		cluster = cv.findClusterForJobConfig("", jc, path, config, jobVolumes)
+	}
+	return cluster
+}
+
+type configResult struct {
+	cluster  string
+	filename string
+	path     string
+}
+
+// dispatchJobs loads the Prow jobs and chooses a cluster in the build farm if possible.
+// The current implementation walks through the Prow Job config files.
+// For each file, it tries to assign all jobs in it to a cluster in the build farm.
+//  - When all the e2e tests are targeting the same cloud provider, we run the test pod on the that cloud provider too.
+//  - When the e2e tests are targeting different cloud providers, or there is no e2e tests at all, we can run the tests
+//    on any cluster in the build farm. Those jobs are used to load balance the workload of clusters in the build farm.
+func dispatchJobs(ctx context.Context, prowJobConfigDir string, maxConcurrency int, config *dispatcher.Config, jobVolumes map[string]float64) error {
+	if config == nil {
+		return fmt.Errorf("config is nil")
+	}
+	// cv stores the volume for each cluster in the build farm
+	cv := &clusterVolume{clusterVolumeMap: map[string]map[string]float64{}, cloudProviders: sets.NewString()}
+	for cloudProvider, v := range config.BuildFarm {
+		cv.cloudProviders.Insert(string(cloudProvider))
+		for cluster := range v {
+			clusterString := string(cluster)
+			cloudProviderString := string(cloudProvider)
+			if _, ok := cv.clusterVolumeMap[cloudProviderString]; !ok {
+				cv.clusterVolumeMap[cloudProviderString] = map[string]float64{}
+			}
+			cv.clusterVolumeMap[cloudProviderString][clusterString] = 0
+		}
+	}
+
+	// no clusters in the build farm
+	if len(cv.clusterVolumeMap) == 0 {
```

Comment:
Shouldn't this be an error?

Reply:
Mainly for backwards compatibility.
Our current config has no `buildFarm` field yet and I still want to run the tool against it.